[Security Solution]Disable the 'Add Endpoint Event Filter' button for non-event and non-…

JiaweiWu opened the pull request:
…endpoint ECS data

## Summary

This PR fixes the issue: https://github.com/elastic/kibana/issues/111846.

Following the suggestions made in the comments, The "Add Endpoint Event Filter" option is now disabled if the ECS data has `event.kind !== 'event'` and `agent.type !== 'endpoint'`

This PR also contains unit tests to verify the 2 cases

### Checklist

Delete any items that are not applicable to this PR.

- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [x] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [x] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [x] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)

JiaweiWu commented:
I added some folks suggested by github, let me know if anyone else should be tagged as reviewer, cheers!

dasansol92 commented:
Hey @JiaweiWu , thanks for taking care of this.

I tried out locally with a real endpoint and found some weird behaviours (see attached gif).

- `Add endpoint event filter` only can be available if event.kind='event' and agent.type='endpoint'. Otherwise it has to be disabled.
- Also noticed that the `Investigate in timeline` option is only available when in the flyout (not from the dropdown menu in list (three dots). Is this expected?
- For some reason, pagination doesn't works, I don't think it's related to this pr, but want to be sure if this is a known bug or not.
- The search bar seems to have a weird behaviour too. When filtering by `not agent.type: "filebeat"` it doesn't seems to apply the filter correctly, am I missing something?

![event filter option disabled](https://user-images.githubusercontent.com/15727784/142602094-3a451ca0-b782-4780-8a7c-a3cdc17bca7c.gif)

Please, let me know if anything is not entirely clear.

dasansol92 commented on x-pack/plugins/security_solution/public/detections/components/take_action_dropdown/index.tsx:
```diff
@@ -87,6 +87,12 @@ export const TakeActionDropdown = React.memo(
+    const isAgentEndpoint = useMemo(() => ecsData?.agent?.type?.includes('endpoint'), [ecsData]);
+
+    const disableEventFilterAction = useMemo(() => {
+      return !isEvent && !isAgentEndpoint;
```
I guess this needs to be an OR condition. So this item is disabled if not is event OR not is endpoint. What it means, it has to be enabled if is an event and is endpoint:

- not (isEvent && isEndpoint) => not isEvent || not isEndpoint

semd replied:
maybe it would be easier if we do `const isEndpointEvent = isEvent && isAgentEndpoint;` and then set the `disabled={!isEndpointEvent}`, too much double negations are a bit confusing.

semd commented:
@dasansol92 the pagination bug has been fixed here https://github.com/elastic/kibana/pull/119065, I am not sure about the other issues you mentioned, you can ask @elastic/security-investigations

JiaweiWu commented:
Thank you @dasansol92 and @semd for taking a look at this PR, I've addressed your comments in a newer commit.

Some other points:
> * Also noticed that the `Investigate in timeline` option is only available when in the flyout (not from the dropdown menu in list (three dots). Is this expected?

The logic for this is in line 166 (`alertsActionItems`) in the `take_action_dropdown` component. Perhaps a separate issue to look into?

> * The search bar seems to have a weird behaviour too. When filtering by `not agent.type: "filebeat"` it doesn't seems to apply the filter correctly, am I missing something?

I tried this in the main branch and tt seems to exhibit the same behaviour. Intuitively I don't think the changes in this PR would alter that functionality

JiaweiWu commented:
@elasticmachine merge upstream

dasansol92 commented on x-pack/plugins/security_solution/public/detections/components/take_action_dropdown/index.tsx:
```diff
@@ -87,6 +87,12 @@ export const TakeActionDropdown = React.memo(
 
+    const isAgentEndpoint = useMemo(() => ecsData?.agent?.type?.includes('endpoint'), [ecsData]);
+
+    const isEndpointEvent = isEvent && isAgentEndpoint;
```
nit: This and the `disableEventFilterAction` could be memoized to avoid unnecessary rerenders.

dasansol92 reviewed (approved):
LGTM! I think you are missing a v8.0 and v8.1 label, right?

JiaweiWu commented:
@elasticmachine merge upstream

JiaweiWu commented:
@elasticmachine merge upstream

kibanamachine commented:
## 💚 Backport successful

| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [8.0](https://github.com/elastic/kibana/pull/119388)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/119388">](https://github.com/elastic/kibana/pull/119388) |
| ✅ |  [7.16](https://github.com/elastic/kibana/pull/119389)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/119389">](https://github.com/elastic/kibana/pull/119389) |

The backport PRs will be merged automatically after passing CI.
